### DISCUSSION TITLE
chaining datasources: first rest response contains references to 1 or more other rest queries

### DISCUSSION THREAD

**Discussion by gsjurseth** [apollographql/apollo-server#5011] [Q&A]
I'm wondering what the right approach would be to chain datasources. I'm building a simple demo application that sits on top of a several microservices with RESTful APIs:

masterdata,
prices
warehouse

prices and warehouse use a uuid-reference to masterdata for all of their data. So far I've written typdefs for masterdata (working: query all and query by name or by id), and typedefs for prices (query all and query by id).

I want the resolvers to chain, but do that the resulting payload needs to have all of the elements (including the nested bits). So what i've done is to add the masterdata datasource to my prices datasource. I then instantiate it there and call initialize on it directly. That way I can add a method to getMasterdataById in my reducer for prices.

This appears to work.... But am I doing this right?

I feel like there's gotta be a better way to have this sorta auto-wire as opposed to me forcing it in my reducer. Am I missing something obvious?

**Comment by glasser** [comment]
I'm not exactly sure I follow what you're asking, but isn't this basically what GraphQL does? It lets you write individual resolvers for each data type and GraphQL execution takes care of attaching all the different types into a response. Or is this specifically about trying to reduce the number of calls to your APIs?

**Reply by gsjurseth** [reply]
Yes, but my issue is that the data being returned doesn't contain all of it. It's not that I'm trying to reduce the API calls, rather, I was hoping that it was possible to wire it up so that it would work natively somehow. Let me try and give a better example.

Imagine the following typedefs: masterdata, warehouse, and prices. And I have datasources for all three of them. The actual API behind masterdata returns a json payload that looks like this:

```json
{
  _id: "c7d6cd32-c3ce-44f5-b959-9c25c92ab324",
  name: "Spoon",
  desc: "a very  nice spoon",
  img: "http://a.great.url.com/images/spoon.jpg"
}
```
An examle from prices may look like this:
```json
{
  _id: "4adcf45d-999c-4bcc-84b3-b5a82726da98",
  md_ref: "c7d6cd32-c3ce-44f5-b959-9c25c92ab324",
  price: 132
}
```

Now i've got datasource classes for each of those that are grabbing those paylaods from the backend. But `md_ref` isn't particular interesting for a query against prices. I'd rather have it use that value and pass it the masterdata function and pull that back automatically. The problem is that I haven't been able to figure out how to do it so what I've done instead is to instantiate the masterdata datasource inside the prices datasource and use it there to flesh out the rest of the json. That way I end up with a json structure that looks like this instead:

```json
{
  _id: "4adcf45d-999c-4bcc-84b3-b5a82726da98",
  md: {
    _id: "c7d6cd32-c3ce-44f5-b959-9c25c92ab324",
    name: "Spoon",
    desc: "a very  nice spoon",
    img: "http://a.great.url.com/images/spoon.jpg"
  },
  price: 132
}
```
Doing that means that when the resolver kicks in (and assuming I've setup the type def to accept md that way) it auto resolves everything appropriately. My typedefs for each of these look like this, btw:

```
  type Masterdata {
    md_id: ID!
    name: String
    desc: String
    img: String
  }

  type Price {
    id: ID
    price: Int
    md: Masterdata
  }
```

Does that make it clearer? It's about the fact that i'm forcing my price datasource to reuse the masterdata datasource to query and flesh out addiontal details. I just feel like there should be a better way.

**Reply by glasser** [reply]
I think the basic idea here is that you define a `Price.md` resolver that uses your masterdata API (passing in the current object's `md_ref` field) to return a Masterdata object.

**Reply by gsjurseth** [reply]
Ah, that  makes sense... Can you point  me to an example? I've followed tutorials such as this one (NOTE: yes, i cheated here and pulled the elements out of the sub-type and put it directly in the main body ... I figured if I were running extra calls anyway I might as well): [https://www.apollographql.com/docs/tutorial/data-source/](url) that uses datasources and reducers that are separate from the resolver itself.

I know i'm asking a lot of questions, but if it helps I can point you at a real datasource here: [https://github.com/gsjurseth/geirs-superfly-istio-demo/blob/master/src/graphql/datasources/catalog.js](url)

The corresponding resolver setup is here: [https://github.com/gsjurseth/geirs-superfly-istio-demo/blob/master/src/graphql/resolvers/index.js](url)

If i'm understanding you correctly then i'd just create the resolver entry end then point it right back at the same datasource. Is that right? So something like this in that case:

```javascript
Price.md: (_, {id}, { dataSources }) =>
          dataSources.MasterdataAPI.getMasterdataById(id)
```

Finally .. Does doing it the way I did it in my datasource break any efficiencies? I ask because one advantage to doing it that way is that I was able to customize what the output looks like. There's an advantage to have the name (which is part of MD) surfaced directly in the Price type (even though I have Price.md). Or perhaps there's a magical way to map that as well?

Thanks for all the help!

**Reply by gsjurseth** [reply]
I tried adding `Price.md` in resolvers and in query as a typedef but it doesn't like the `.`. So i'm not sure how to wire this up.

**Reply by glasser** [reply]
That there is the general pattern for fields that look up other objects in databases.

I don't think I can provide guidance on your "doesn't like the `.`" issue without a full reproduction.

**Comment by gsjurseth** [comment]
I just mean that it wouldn't let me have a resolver and a typedef with a
dot in the name. So `Price.md` wasn't an allowed name.

/geir

On Thu, Mar 11, 2021 at 9:20 PM David Glasser ***@***.***>
wrote:

> That there is the general pattern for fields that look up other objects in
> databases.
>
> I don't think I can provide guidance on your "doesn't like the ." issue
> without a full reproduction.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apollographql/apollo-server/discussions/5011#discussioncomment-471291>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABL5LEOENFVVQQJPHFE2ACDTDEJWDANCNFSM4Y2DWNSQ>
> .
>


-- 
Geir Sjurseth |  Tech Strategist EMEA (Apigee) |  ***@***.*** |  +46
734 48 67 14

**Reply by glasser** [reply]
I mean the md field on the Price type.
